numpy/core/src/umath/ufunc_type_resolution.c

```diff
@@ -1508,6 +1507,22 @@ ufunc_loop_matches(PyUFuncObject *self,
             tmp = dtypes[i];
             Py_INCREF(tmp);
         }
+        else if (types[i] == PyArray_TYPE(op[i])) {
+            /*
+             * Don't bother doing casting checks if the type numbers exactly
+             * match.
```

Comment:
It might be safer to do this check outside the loop so that it only checks the first element, possibly special cased for timedelta/datetime. I can see this check going wrong for multiple input arguments.